Issue #474: Make default scoring rules functions rather than stored data sets

nikosbosse opened the pull request:
<!-- Thanks for opening this pull request! Below we have provided a suggested template for PRs to this repository and a checklist to complete before opening a PR -->
 
## Description

This PR fixes #474.

It 
- replaces the previous data sets (`metrics_point`, `metrics_binary`, `metrics_quantile` and `metrics_sample`) with functions that return a list with functions used as default scoring rules (`rules_point()`, `rules_binary()`, `rules_quantile()` and `rules_sample()`)
- introduces a new helper function, `select_rules()` to implement basic functionality to select or exclude functions from the default list in a call to `rules_*()`
- updates existing tests and adds new tests for the new functions
- replaces a previous `\(...)` with `function(...)` to make sure everything can be run on `R` 3.6
- updates the NEWS file

Further considerations: 
- This PR does not yet resolve all naming consistencies (i.e. "metrics" and "rules") see https://github.com/epiforecasts/scoringutils/issues/476. I suggest addressing this separately. 
- The PR replaces existing functionality. We talked about additional features such as making scoring rules modular and composable, e.g. by creating a function like `rules_wis()` that includes all WIS rules.
- In https://github.com/epiforecasts/scoringutils/issues/365 and https://github.com/epiforecasts/scoringutils/issues/415 we discuss additional ways to document default scoring rules. 
  - From previous versions there still is a data object, `metrics`, which is a data.table of explanations. I think we should ultimately get rid of this
  - We previously talked about a `print()` method for the scoring rules with further explanations. I'm not convinced anymore we really need this. The documentation for `rules_*()` links to the documentation for the additional functions, which should have all the explanations the user needs. In addition, users should be able to find everything in the vignettes. Are we happy with this? If so I suggest deleting the stored `metrics()` object.
  - We could, however, have a `print()` method that simply adds a sentence "the following scores will be computed: 'names of scores'" before printing the list. This could be nice, but maybe not necessary

## Checklist

- [x] My PR is based on a package issue and I have explicitly linked it.
- [x] I have included the target issue or issues in the PR title as follows: *issue-number*: PR title
- [x] I have tested my changes locally.
- [x] I have added or updated unit tests where necessary.
- [x] I have updated the documentation if required.
- [x] I have built the package locally and run rebuilt docs using roxygen2.
- [x] My code follows the established coding standards and I have run `lintr::lint_package()` to check for style issues introduced by my changes. 
- [x] I have added a news item linked to this PR.
- [x] I have reviewed CI checks for this PR and addressed them as far as I am able.

<!-- Thanks again for this PR - @scoringutils dev team -->

seabbs commented:
> Are we happy with this? If so I suggest deleting the stored metrics() object.

Yes agree. We should do this in its own PR though so that we are sure we have the documentation in place to replace it. 

> We could, however, have a print() method that simply adds a sentence "the following scores will be computed: 'names of scores'" before printing the list. This could be nice, but maybe not necessary

Yes this is a nice idea and feels like its own issue.

seabbs commented on R/default-scoring-rules.R:
```diff
@@ -0,0 +1,155 @@
+#' @title Select Scoring Rules From A List of Possibilities
+#' @description Helper function to return only the scoring rules selected by
```
If this is a helper function for users why isn't it exported?

nikosbosse replied:
I thought of it as an internal helper function that's used to reduce the amount of duplicated code. It's just the internal implementation of the exclude/include functionality, so not sure users would ever need it. Do you think there is a use case?

seabbs replied:
I think when it comes to combining and creating custom lists there is likely a use case and personally would prefer to call the function containing the defaults and then pipe into the selection function I think

nikosbosse replied:
Good with me. Are you fine with keeping both options? I.e. allowing both
```
rules_quantile() |>
  select_rules("wis")
```
and
```
rules_quantile("wis")
```

Or do you think the `select` and `exclude` arguments should be removed from `rules_*()`?

nikosbosse replied:
I think I like having both because it allows you to 
a) compose your scores as you want
b) do something like `score(data, rules_quantile("wis")` in a concise way

seabbs replied:
I'm fine with having both for sure.

seabbs commented on R/default-scoring-rules.R:
```diff
@@ -0,0 +1,155 @@
+#' the user from a list of possible scoring rules.
+#' @param possibilities A list of scoring rules.
+#' @param select A character vector of scoring rules to select from the list.
+#' If `select = all` (the default), all possible scoring rules are returned.
```
It looks like the default is being missing (and then this is handled as all).

nikosbosse replied:
Good catch, I updated the default.

seabbs commented on R/default-scoring-rules.R:
```diff
@@ -0,0 +1,155 @@
+#' @title Select Scoring Rules From A List of Possibilities
+#' @description Helper function to return only the scoring rules selected by
+#' the user from a list of possible scoring rules.
+#' @param possibilities A list of scoring rules.
```
This seems like a weird name for this.

seabbs commented on R/default-scoring-rules.R:
```diff
@@ -0,0 +1,155 @@
+#'   possibilities = rules_binary(),
+#'   select = "brier_score"
+#' )
+select_rules <- function(possibilities, select, exclude = NULL) {
```
having one argument be select and one being exclude seems quite clunky naming-wise. Why not include and exclude?

nikosbosse replied:
makes sense, I didn't think of that. I'll update the names

nikosbosse replied:
hm. thinking of it, "include" is not exclusive - while select is. I.e. if you say "I select option A" it means "you pick only A". But if you include A, then others may be present as well. What do you think?

seabbs replied:
yeah okay I think you are right and that the current naming makes sense.

seabbs commented on R/default-scoring-rules.R:
```diff
@@ -0,0 +1,155 @@
+#' - "dispersion" = [dispersion()]
+#' - "bias" = [bias_quantile()]
+#' - "coverage_50" = [interval_coverage_quantile()]
+#' - "coverage_90" = \(...) \{
```
internal uses function() - perhaps documenting this in a slightly more human readable way might be nice vs just reproducing the code?

seabbs commented on R/score.R:
```diff
@@ -76,7 +76,7 @@ score.default <- function(data, ...) {
 #' @rdname score
 #' @export
-score.forecast_binary <- function(data, metrics = metrics_binary, ...) {
+score.forecast_binary <- function(data, metrics = rules_binary(), ...) {
```
do these need to be scoringutils::rules_binary etc to work when scoringutils isn't loaded?

nikosbosse replied:
The following works fine even with `scoringutils` not loaded:

`scoringutils::score(data.frame(predicted = 1:10, observed = 1:10, id = 1:10))`

So I think it should be fine

seabbs replied:
hmm interesting. well seems like magic to me

seabbs commented on tests/testthat/test-default-scoring-rules.R:
```diff
@@ -0,0 +1,71 @@
+  )
+
+  expect_equal(
+    names(scoringutils:::select_rules(rules_point(), select = "ape")),
```
this seems like quite a nice workflow for users?

nikosbosse replied:
but they can just do `rules_point(select = "ape")` instead, which is more concise. 

I can see a case once you start combining several lists, e.g. `select_rules(a(), b(), select = c("c", "d").

seabbs replied:
> I can see a use case once you start combining several lists,

This would be the argument and the push back on it being more concise is that yes that is true but then each function is doing multiple things which can confuse readers. I don't have an extremely strong opinion but think if should be exported and if keeping as is then should at least make use of match.arg or the rlang version) to get some functionality.

seabbs reviewed (commented):
This looks all good in the main. I have a slight query about the workflow with select_rules and it being internal vs external. See the specific comments.

nikosbosse commented on R/default-scoring-rules.R:
```diff
@@ -0,0 +1,155 @@
+  selected <- select_rules(all, select, exclude)
+  return(selected)
+}
+
```
`select_rules()` calls `assert_subset()`, which is doing the equivalent of `match.arg()`. But that's not quite what you mean, is it?

seabbs replied:
ah okay - great

nikosbosse commented on R/default-scoring-rules.R:
```diff
@@ -0,0 +1,155 @@
+  assert_list(possibilities)
+  allowed <- names(possibilities)
+
+  if (select == "all" && is.null(exclude)) {
```
Fine with me. In other functions we usually have `NULL` as default instead of missing though

seabbs replied:
have it be null then

nikosbosse replied:
Changed it

nikosbosse commented:
Latest updates: 
- exported `select_rules()`, but also left the arguments `select` and `exclude` in `rules_*()` in place
- added more explanation to the `coverage_90` function in `rules_quantile()` and changed `\(...)` to `function(...)`
- fixed a random typo in the documentation of `run_safely()`
- changed pkgdown keywords to "metric", although as mentioned previously we need to rethink those at some point
- changed `select = all` to `select = NULL` as the default and updated tests accordingly

seabbs reviewed (commented):
LGTM. This seems really slick vs the old version for some reason.

seabbs reviewed (approved):
LGTM. This seems really slick vs the old version for some reason.

nikosbosse commented:
Thanks a lot for reviewing!
